quickwit-search/src/client_pool/search_client_pool.rs

```diff
@@ -32,26 +32,58 @@ use tracing::*;
 
 use quickwit_cluster::cluster::Cluster;
 
-use crate::client::{create_search_service_client, WrappedSearchServiceClient};
+use crate::client::create_search_service_client;
 use crate::client_pool::{ClientPool, Job};
 use crate::rendezvous_hasher::{sort_by_rendez_vous_hash, Node};
 use crate::swim_addr_to_grpc_addr;
+use crate::SearchServiceClient;
 
 /// Search client pool implementation.
 #[derive(Clone)]
 pub struct SearchClientPool {
     /// Search clients.
     /// A hash map with gRPC's SocketAddr as the key and SearchServiceClient as the value.
     /// It is not the cluster listen address.
-    pub clients: Arc<RwLock<HashMap<SocketAddr, WrappedSearchServiceClient>>>,
+    pub clients: Arc<RwLock<HashMap<SocketAddr, SearchServiceClient>>>,
 }
 
 impl SearchClientPool {
+    #[cfg(test)]
+    pub async fn from_mocks(
+        mock_services: Vec<Arc<dyn crate::SearchService>>,
+    ) -> anyhow::Result<Self> {
+        let mut mock_clients = HashMap::new();
+        for (mock_ord, mock_service) in mock_services.into_iter().enumerate() {
+            let grpc_addr: SocketAddr =
+                format!("127.0.0.1:{}", 10000 + mock_ord as u16 * 10).parse()?;
+            let mock_client = SearchServiceClient::from_service(mock_service, grpc_addr);
+            mock_clients.insert(grpc_addr, mock_client);
+        }
+
+        Ok(SearchClientPool {
+            clients: Arc::new(RwLock::new(mock_clients)),
+        })
+    }
+
     /// Create a search client pool given a cluster.
     /// When a client pool is created, the thread that monitors cluster members
     /// will be started at the same time.
-    pub async fn new(cluster: Arc<Cluster>) -> anyhow::Result<Self> {
-        let clients = HashMap::new();
+    pub async fn new(cluster: Arc<dyn Cluster>) -> anyhow::Result<Self> {
+        let mut clients = HashMap::new();
+
+        // Initialize the client pool with members of the cluster.
+        for member in cluster.members() {
+            let grpc_addr = swim_addr_to_grpc_addr(member.listen_addr);
+            match create_search_service_client(grpc_addr).await {
+                Ok(client) => {
+                    debug!(grpc_addr=?grpc_addr, "Add a new client to connect to the members of the cluster.");
+                    clients.insert(grpc_addr, client);
+                }
+                Err(err) => {
+                    error!(grpc_addr=?grpc_addr, err=?err, "Failed to create search client.")
+                }
```

Comment:
This may not be the best example, but we should think about feature flags, where to add them and how to control them.

E.g. here I could image a setting that optionally ignores errors
(should setting up a cluster fail, because the client creation for one member failed?)

Reply:
Indeed. I hadn't thought about control by feature flags yet.
I guess I need to think about what I want to control with feature flags.

Reply:
I don't think this sounds like a feature flag setting, but more like something that is dynamic.

It is per-case discussion and it is difficult to judge, but in this specific case, I think one default behavior should fit
everyone 's liking.

Reply:
I think it would be better to remove the node and fail if the error rate is too high.

After a certain cluster size (what's the target cluster size btw) and client failure rate too many requests could fail.

Reply:
Absolutely. Although keep in mind this happens if the node is considered "healthy" by the SWIM cluster, but we somehow cannot connect to it.

(We could see that if some firewall refuses TCP connection but accepts UDP connection for instance... But it is not a super pathological case)